src/twisted/protocols/_sni.py

```diff
@@ -0,0 +1,281 @@
+        self.clock = clock
+
+    def listen(self, factory: IProtocolFactory) -> Deferred[IListeningPort]:
+        return self.endpoint.listen(
+            TLSMemoryBIOFactory(self.contextFactory, False, factory, clock=self.clock)
+        )
+
+
+def _getSubjectAltNames(c: Certificate) -> List[str]:
+    """
+    Get all the DNSName SANs for a given certificate.
+    """
+    return [
+        value
+        for extension in load_pem_x509_certificate(c.dumpPEM()).extensions
+        if extension.oid == ExtensionOID.SUBJECT_ALTERNATIVE_NAME
+        for value in extension.value.get_values_for_type(DNSName)
+    ]
+
+
+def autoReloadingDirectoryOfPEMs(
+    path: FilePath[str],
+) -> Callable[[bytes | None], Context | None]:
+    """
+    Construct a callable that can look up a HTTPS certificate based on their
+    DNS names, by inspecting a directory full of PEM objects.  When
+    encountering a lookup failure, the directory will be reloaded, so that if
+    new certificates are added they will be picked up.
+    """
+    # TODO: some flaws with this approach
+
+    """
+        1. too much re-scanning; re-reading full file contents for every single
+           certificate even if only one has changed.  a mtime/length cache
+           would be a good place to start with this
+
+        2. too trusting; we get a network request for a billion certificate
+           names per second, we go ahead and do a bunch of work every single
+           time (and, see point 1, re-scan and re-parse every single file)
+
+        3. not *enough* re-scanning on the happy path; if certificates go
+           stale, we just let them sit there until we get an unknown hostname
+
+        4. we don't look at notAfter/notBefore, so if we find multiple certs,
+           we may end up using the wrong one
+
+    """
+
+    certMap: dict[str, CertificateOptions]
+
+    def doReload() -> None:
+        nonlocal certMap
+        certMap = PEMObjects.fromDirectory(path).inferDomainMapping()
+
+    def lookup(name: bytes | None, shouldReload: bool = True) -> Context | None:
+        name = next(iter(certMap.keys()), "").encode() if name in (None, b"") else name
+        assert name is not None
+        if (options := certMap.get(name.decode())) is not None:
+            return options.getContext()
+        if not shouldReload:
+            return None
+        msg = "could not find domain {name}, re-loading {path}"
+        log.warn(msg, name=name, path=path)
+        doReload()
+        return lookup(name, False)
+
+    doReload()
+    return lookup
+
+
+@dataclass
+class PEMObjects:
+    """
+    A collection of objects loaded from a PEM encoded file.
+    """
+
+    certificates: List[tuple[FilePath[str], Certificate]]
+    keyPairs: List[tuple[FilePath[str], KeyPair]]
+
+    @classmethod
+    def fromDirectory(cls, directory: FilePath[str]) -> PEMObjects:
+        """
+        Load a single PEMObjects from all the PEMs in a big directory.
```

Comment:
Should say something like "recursively walk a directory looking for files with a `.pem` extension".

Also, should this ignore hidden files? Right now it would load a file with a basename of `.pem`.